// ngx-show-hide-password/bundles/ngx-show-hide-password.umd.js
(function (global, factory) {
    typeof exports === 'object' && typeof module !== 'undefined' ? factory(exports, require('@angular/core'), require('rxjs'), require('@fortawesome/free-solid-svg-icons'), require('@angular/common'), require('@angular/forms'), require('@fortawesome/angular-fontawesome')) :
    typeof define === 'function' && define.amd ? define('ngx-show-hide-password', ['exports', '@angular/core', 'rxjs', '@fortawesome/free-solid-svg-icons', '@angular/common', '@angular/forms', '@fortawesome/angular-fontawesome'], factory) :
    (global = global || self, factory(global['ngx-show-hide-password'] = {}, global.ng.core, global.rxjs, global.freeSolidSvgIcons, global.ng.common, global.ng.forms, global.angularFontawesome));
}(this, function (exports, core, rxjs, freeSolidSvgIcons, common, forms, angularFontawesome) { 'use strict';

    /*! *****************************************************************************
    Copyright (c) Microsoft Corporation. All rights reserved.
    Licensed under the Apache License, Version 2.0 (the "License"); you may not use
    this file except in compliance with the License. You may obtain a copy of the
    License at http://www.apache.org/licenses/LICENSE-2.0

    THIS CODE IS PROVIDED ON AN *AS IS* BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
    KIND, EITHER EXPRESS OR IMPLIED, INCLUDING WITHOUT LIMITATION ANY IMPLIED
    WARRANTIES OR CONDITIONS OF TITLE, FITNESS FOR A PARTICULAR PURPOSE,
    MERCHANTABLITY OR NON-INFRINGEMENT.

    See the Apache Version 2.0 License for specific language governing permissions
    and limitations under the License.
    ***************************************************************************** */

    var __assign = function() {
        __assign = Object.assign || function __assign(t) {
            for (var s, i = 1, n = arguments.length; i < n; i++) {
                s = arguments[i];
                for (var p in s) if (Object.prototype.hasOwnProperty.call(s, p)) t[p] = s[p];
            }
            return t;
        };
        return __assign.apply(this, arguments);
    };

    /**
     * @fileoverview added by tsickle
     * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
     */
    var ShowHideService = /** @class */ (function () {
        function ShowHideService() {
            this.states = [];
        }
        /**
         * @private
         * @param {?} id
         * @return {?}
         */
        ShowHideService.prototype.getIO = /**
         * @private
         * @param {?} id
         * @return {?}
         */
        function (id) {
            /** @type {?} */
            var io = this.states.find((/**
             * @param {?} o
             * @return {?}
             */
            function (o) { return o.id === id; }));
            if (!io && id) {
                io = this.init(id);
            }
            return io;
        };
        /**
         * @private
         * @param {?} id
         * @return {?}
         */
        ShowHideService.prototype.init = /**
         * @private
         * @param {?} id
         * @return {?}
         */
        function (id) {
            /** @type {?} */
            var subject = new rxjs.ReplaySubject(1);
            /** @type {?} */
            var io = {
                id: id,
                show: false,
                subject: subject
            };
            this.states.push(io);
            return io;
        };
        /**
         * @private
         * @param {?} io
         * @param {?} show
         * @return {?}
         */
        ShowHideService.prototype.saveAndProadcast = /**
         * @private
         * @param {?} io
         * @param {?} show
         * @return {?}
         */
        function (io, show) {
            io.show = show;
            io.subject.next(io.show);
        };
        /**
         * @param {?} id
         * @return {?}
         */
        ShowHideService.prototype.getObservable = /**
         * @param {?} id
         * @return {?}
         */
        function (id) {
            return this.getIO(id).subject;
        };
        /**
         * @param {?} id
         * @param {?} show
         * @return {?}
         */
        ShowHideService.prototype.setShow = /**
         * @param {?} id
         * @param {?} show
         * @return {?}
         */
        function (id, show) {
            this.saveAndProadcast(this.getIO(id), show);
        };
        /**
         * @param {?} id
         * @return {?}
         */
        ShowHideService.prototype.toggleShow = /**
         * @param {?} id
         * @return {?}
         */
        function (id) {
            /** @type {?} */
            var io = this.getIO(id);
            this.saveAndProadcast(io, !io.show);
        };
        ShowHideService.decorators = [
            { type: core.Injectable, args: [{
                        providedIn: 'root'
                    },] }
        ];
        /** @nocollapse */
        ShowHideService.ctorParameters = function () { return []; };
        /** @nocollapse */ ShowHideService.ngInjectableDef = core.ɵɵdefineInjectable({ factory: function ShowHideService_Factory() { return new ShowHideService(); }, token: ShowHideService, providedIn: "root" });
        return ShowHideService;
    }());

    /**
     * @fileoverview added by tsickle
     * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
     */
    var ShowHideTriggerDirective = /** @class */ (function () {
        function ShowHideTriggerDirective(service, errorHandler) {
            this.service = service;
            this.errorHandler = errorHandler;
        }
        /**
         * @return {?}
         */
        ShowHideTriggerDirective.prototype.onClick = /**
         * @return {?}
         */
        function () {
            if (this.showHideTrigger) {
                this.service.toggleShow(this.showHideTrigger);
            }
            else {
                this.errorHandler.handleError(new Error("No input id found. Please read the docs!"));
            }
        };
        ShowHideTriggerDirective.decorators = [
            { type: core.Directive, args: [{
                        selector: '[showHideTrigger]'
                    },] }
        ];
        /** @nocollapse */
        ShowHideTriggerDirective.ctorParameters = function () { return [
            { type: ShowHideService },
            { type: core.ErrorHandler }
        ]; };
        ShowHideTriggerDirective.propDecorators = {
            showHideTrigger: [{ type: core.Input }],
            onClick: [{ type: core.HostListener, args: ['click',] }]
        };
        return ShowHideTriggerDirective;
    }());

    /**
     * @fileoverview added by tsickle
     * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
     */
    var ShowHideStatusDirective = /** @class */ (function () {
        function ShowHideStatusDirective(service, el, renderer, errorHandler) {
            this.service = service;
            this.el = el;
            this.renderer = renderer;
            this.errorHandler = errorHandler;
        }
        Object.defineProperty(ShowHideStatusDirective.prototype, "showHideStatus", {
            set: /**
             * @param {?} config
             * @return {?}
             */
            function (config) {
                this.init(config);
            },
            enumerable: true,
            configurable: true
        });
        /**
         * @private
         * @param {?} config
         * @return {?}
         */
        ShowHideStatusDirective.prototype.init = /**
         * @private
         * @param {?} config
         * @return {?}
         */
        function (config) {
            var _this = this;
            /** @type {?} */
            var defaultConfig = {
                show: 'visibility',
                hide: 'visibility_off',
                materialIcon: false,
                id: null
            };
            this.config = __assign({}, defaultConfig, config);
            if (this.config.id) {
                this.subscription = this.service.getObservable(this.config.id).subscribe((/**
                 * @param {?} show
                 * @return {?}
                 */
                function (show) { return _this.updateStatus(show); }));
            }
            else {
                this.errorHandler.handleError(new Error("No input id found. Please read the docs!"));
            }
        };
        /**
         * @private
         * @param {?} show
         * @return {?}
         */
        ShowHideStatusDirective.prototype.updateStatus = /**
         * @private
         * @param {?} show
         * @return {?}
         */
        function (show) {
            if (this.config.materialIcon) {
                this.renderer.setProperty(this.el.nativeElement, 'innerHTML', show ? this.config.hide : this.config.show);
            }
            else {
                this.renderer.removeClass(this.el.nativeElement, !show ? this.config.hide : this.config.show);
                this.renderer.addClass(this.el.nativeElement, show ? this.config.hide : this.config.show);
            }
        };
        /**
         * @return {?}
         */
        ShowHideStatusDirective.prototype.ngOnDestroy = /**
         * @return {?}
         */
        function () {
            if (this.subscription) {
                this.subscription.unsubscribe();
            }
        };
        ShowHideStatusDirective.decorators = [
            { type: core.Directive, args: [{
                        selector: '[showHideStatus]'
                    },] }
        ];
        /** @nocollapse */
        ShowHideStatusDirective.ctorParameters = function () { return [
            { type: ShowHideService },
            { type: core.ElementRef },
            { type: core.Renderer2 },
            { type: core.ErrorHandler }
        ]; };
        ShowHideStatusDirective.propDecorators = {
            showHideStatus: [{ type: core.Input }]
        };
        return ShowHideStatusDirective;
    }());

    /**
     * @fileoverview added by tsickle
     * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
     */
    var ShowHideInputDirective = /** @class */ (function () {
        function ShowHideInputDirective(service, el, renderer) {
            this.service = service;
            this.el = el;
            this.renderer = renderer;
            this.id = this.el.nativeElement.getAttribute('id');
            if (!this.id) {
                throw new Error("No input id found. Please read the docs!");
            }
            this.service.setShow(this.id, this.el.nativeElement.type !== 'password');
        }
        /**
         * @return {?}
         */
        ShowHideInputDirective.prototype.ngOnInit = /**
         * @return {?}
         */
        function () {
            var _this = this;
            this.subscription = this.service
                .getObservable(this.id)
                .subscribe((/**
             * @param {?} show
             * @return {?}
             */
            function (show) { return _this.renderer.setAttribute(_this.el.nativeElement, 'type', show ? 'text' : 'password'); }));
        };
        /**
         * @return {?}
         */
        ShowHideInputDirective.prototype.ngOnDestroy = /**
         * @return {?}
         */
        function () {
            if (this.subscription) {
                this.subscription.unsubscribe();
            }
        };
        ShowHideInputDirective.decorators = [
            { type: core.Directive, args: [{
                        selector: 'input[showHideInput]'
                    },] }
        ];
        /** @nocollapse */
        ShowHideInputDirective.ctorParameters = function () { return [
            { type: ShowHideService },
            { type: core.ElementRef },
            { type: core.Renderer2 }
        ]; };
        return ShowHideInputDirective;
    }());

    /**
     * @fileoverview added by tsickle
     * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
     */
    /** @enum {string} */
    var BtnStyle = {
        Primary: 'primary',
        Secondary: 'secondary',
        Success: 'success',
        Danger: 'danger',
        Warning: 'warning',
        Info: 'info',
        Dark: 'dark',
        Light: 'light',
    };
    /**
     * Add a split input button to password or text input. Toggles input type between "text" and "password".
     *
     * \@example
     * <show-hide-password size="sm|lg">
     * <input type="password" name=... />
     * </show-hide-password>
     */
    var ShowHidePasswordComponent = /** @class */ (function () {
        function ShowHidePasswordComponent(service, elem, renderer) {
            this.service = service;
            this.elem = elem;
            this.renderer = renderer;
            this.btnStyle = BtnStyle.Secondary;
            this.btnOutline = true;
            this.faEye = freeSolidSvgIcons.faEye;
            this.faEyeSlash = freeSolidSvgIcons.faEyeSlash;
        }
        /**
         * @return {?}
         */
        ShowHidePasswordComponent.prototype.ngOnInit = /**
         * @return {?}
         */
        function () {
            var _this = this;
            this.input = this.elem.nativeElement.querySelector('input');
            if (!this.input) {
                throw new Error("No input element found. Please read the docs!");
            }
            this.id = this.input.getAttribute('id');
            if (!this.id) {
                this.id = 'showHideInput' + Math.round(Math.random() * 100000);
                this.renderer.setAttribute(this.input, 'id', this.id);
            }
            this.renderer.addClass(this.elem.nativeElement, 'input-group');
            if (this.size === 'sm') {
                this.renderer.addClass(this.elem.nativeElement, 'input-group-sm');
            }
            else if (this.size === 'lg') {
                this.renderer.addClass(this.elem.nativeElement, 'input-group-lg');
            }
            this.isHidden = this.input.type === 'password';
            this.renderer.addClass(this.input, 'form-control'); // just to be sure
            this.service.setShow(this.id, this.input.type !== 'password');
            this.subscription = this.service.getObservable(this.id).subscribe((/**
             * @param {?} show
             * @return {?}
             */
            function (show) {
                _this.isHidden = !show;
                _this.renderer.setAttribute(_this.input, 'type', show ? 'text' : 'password');
            }));
        };
        /**
         * @return {?}
         */
        ShowHidePasswordComponent.prototype.ngOnDestroy = /**
         * @return {?}
         */
        function () {
            if (this.subscription) {
                this.subscription.unsubscribe();
            }
        };
        ShowHidePasswordComponent.decorators = [
            { type: core.Component, args: [{
                        selector: 'show-hide-password',
                        changeDetection: core.ChangeDetectionStrategy.OnPush,
                        template: "\n  <ng-content></ng-content>\n  <div class=\"input-group-append ngx-show-hide-password\">\n    <button class=\"btn\" [ngClass]=\"(btnOutline ? 'btn-outline-' + btnStyle : 'btn-' + btnStyle)\"\n      type=\"button\" [showHideTrigger]=\"id\">\n      <fa-icon [fixedWidth]=\"true\" size=\"lg\" [icon]=\"(isHidden ? faEye : faEyeSlash)\"\n        [showHideStatus]=\"{id: id}\"></fa-icon>\n    </button>\n  </div>\n"
                    }] }
        ];
        /** @nocollapse */
        ShowHidePasswordComponent.ctorParameters = function () { return [
            { type: ShowHideService },
            { type: core.ElementRef },
            { type: core.Renderer2 }
        ]; };
        ShowHidePasswordComponent.propDecorators = {
            btnStyle: [{ type: core.Input }],
            btnOutline: [{ type: core.Input }],
            size: [{ type: core.Input }]
        };
        return ShowHidePasswordComponent;
    }());

    /**
     * @fileoverview added by tsickle
     * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
     */
    var ShowHidePasswordModule = /** @class */ (function () {
        function ShowHidePasswordModule() {
        }
        ShowHidePasswordModule.decorators = [
            { type: core.NgModule, args: [{
                        imports: [common.CommonModule, forms.FormsModule, angularFontawesome.FontAwesomeModule],
                        exports: [ShowHidePasswordComponent, ShowHideTriggerDirective, ShowHideInputDirective, ShowHideStatusDirective],
                        declarations: [ShowHidePasswordComponent, ShowHideTriggerDirective, ShowHideInputDirective, ShowHideStatusDirective]
                    },] }
        ];
        return ShowHidePasswordModule;
    }());

    exports.BtnStyle = BtnStyle;
    exports.ShowHideInputDirective = ShowHideInputDirective;
    exports.ShowHidePasswordComponent = ShowHidePasswordComponent;
    exports.ShowHidePasswordModule = ShowHidePasswordModule;
    exports.ShowHideService = ShowHideService;
    exports.ShowHideStatusDirective = ShowHideStatusDirective;
    exports.ShowHideTriggerDirective = ShowHideTriggerDirective;

    Object.defineProperty(exports, '__esModule', { value: true });

}));
